signer/src/dkg/verification.rs

```diff
@@ -0,0 +1,804 @@
+    EndState(Box<State>),
+
+    /// The state machine has reached the message limit for the given message
+    /// type.
+    #[error("the state machine has reached the message limit for message type {message_type:?} (expected: {expected}, actual: {actual})")]
+    MessageLimitExceeded {
+        /// The type of message that has reached the limit.
+        message_type: WstsNetMessageType,
+        /// The expected number of messages of this type that should be
+        /// received.
+        expected: u32,
+        /// The actual number of messages of this type that have been received.
+        actual: u32,
+    },
+
+    /// The FROST coordinator returned an error.
+    #[error("a signing error occurred: {0}")]
+    SigningFailure(Box<SignError>),
+
+    /// The FROST coordinator returned an unexpected result.
+    #[error("unexpected WSTS result: {0:?}")]
+    UnexpectedWstsResult(Box<OperationResult>),
+
+    /// An error occurred in the WSTS coordinator.
+    #[error("WSTS coordinator error: {0}")]
+    Coordinator(Box<dyn std::error::Error + 'static + Send + Sync>),
+}
+
+/// Represents the state of a DKG verification.
+#[derive(Debug, Clone)]
+pub enum State {
+    /// The DKG verification has been created but not yet been used to process
+    /// any messages.
+    Idle,
+    /// The DKG verification signing round has begun but has not yet finalized.
+    Signing,
+    /// The DKG verification has completed successfully, and the resulting
+    /// signature is stored.
+    Success(TaprootSignature),
+    /// The DKG verification has failed due an an error or unexpected result
+    /// from the FROST coordinator.
+    Error,
+    /// The DKG verification has expired and can no longer be used.
+    Expired,
+}
+
+impl std::fmt::Display for State {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            State::Idle => write!(f, "idle"),
+            State::Signing => write!(f, "signing"),
+            State::Success(_) => write!(f, "success"),
+            State::Error => write!(f, "error"),
+            State::Expired => write!(f, "expired"),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct QueuedMessage {
+    processed: bool,
+    message: wsts::net::Message,
+}
+
+impl QueuedMessage {
+    fn new(message: wsts::net::Message) -> Self {
+        Self { processed: false, message }
+    }
+
+    fn mark_processed(&mut self) {
+        self.processed = true;
+    }
+}
+
+/// Represents the state of a DKG verification. This implementation is designed
+/// to be able to span across multiple Bitcoin blocks, being limited by
+/// wall-clock time instead of Bitcoin block cadence.
+///
+/// TODO: The signer currently stops messages from being processed if the
+/// received message doesn't match the current chain tip, so that needs to be
+/// changed for cross-block functionality to work.
+#[allow(dead_code)]
+#[derive(Debug)]
+pub struct StateMachine<TError> {
```

Comment:
Do we need the error template? It's not like we use this with something other than the crate `Error`

Reply:
I'd prefer to keep it because it keeps this file clean and concise wrt errors, and the supers can use a type alias if it's too tedious to write `StateMachine<Error>`.

Reply:
But I'll look at it tomorrow, it might be that the reason I needed it went away with some other changes I made recently

Reply:
Yeah it's not needed, [5f8a13e](https://github.com/stacks-network/sbtc/pull/1324/commits/5f8a13e2890cdf94b834518f78e79a6eb9949d55)